File: project/backend-services/rule-management-backend/infrastructure/persistence/postgres/src/main/java/pt/sensae/services/rule/management/backend/infrastructure/persistence/postgres/mapper/RuleScenarioMapper.java
package pt.sensae.services.rule.management.backend.infrastructure.persistence.postgres.mapper;

import pt.sensae.services.rule.management.backend.domain.*;
import pt.sensae.services.rule.management.backend.infrastructure.persistence.postgres.model.RuleScenarioPostgres;

import java.util.Arrays;
import java.util.UUID;
import java.util.stream.Collectors;

public class RuleScenarioMapper {

    public static RuleScenario postgresToDomain(RuleScenarioPostgres postgres) {
        var scenarioId = RuleScenarioId.of(postgres.scenarioId);
        var content = RuleScenarioContent.of(postgres.scenarioContent);
        var domains = RuleScenarioOwners.of(Arrays.stream(postgres.scenarioOwners)
                .map(UUID::fromString)
                .collect(Collectors.toSet()));
        return new RuleScenario(scenarioId, content, domains, new RuleApplied(postgres.applied), new RuleDeleted(postgres.deleted));
    }

    public static RuleScenarioPostgres domainToPostgres(RuleScenario domain) {
        var ruleScenario = new RuleScenarioPostgres();
        ruleScenario.scenarioId = domain.id().getValue();
        ruleScenario.scenarioContent = domain.content().value();
        ruleScenario.scenarioOwners = domain.owners()
                .domains()
                .stream()
                .map(UUID::toString)
                .toList()
                .toArray(new String[0]);
        ruleScenario.applied = domain.isApplied().value();
        return ruleScenario;
    }
}

File: project/frontend-services/libs/rule-management/mapper/src/lib/mapper/RuleScenarioMapper.ts
import {RuleScenarioIdMapper} from './RuleScenarioIdMapper';
import {IsDraft, RuleScenario, RuleScenarioContent} from "@frontend-services/rule-management/model";
import {RuleScenarioDTO, RuleScenarioInputDTO} from "@frontend-services/rule-management/dto";

export class RuleScenarioMapper {
  static dtoToModel(dto: RuleScenarioDTO): RuleScenario {
    const ruleScenarioId = RuleScenarioIdMapper.dtoToModel(dto.id);
    const content = new RuleScenarioContent(dto.content);
    const isDraft = new IsDraft(!dto.applied);
    return new RuleScenario(ruleScenarioId, content, isDraft);
  }

  static modelToDto(model: RuleScenario): RuleScenarioInputDTO {
    return {id: {value: model.id.value}, content: model.content.value};
  }
}
